airbyte-integrations/connectors/source-google-directory/source_google_directory/api.py

```diff
@@ -0,0 +1,140 @@
+
+
+class API:
+    def __init__(self, credentials_json: str, credentials_pickle: str):
+        self._creds = None
+        self._credentials_json = credentials_json
+        self._credentials_pickle = credentials_pickle
+
+    def _load_cached_creds(self):
+        if os.path.exists(self._credentials_pickle):
+            with open(self._credentials_pickle, "rb") as token:
+                self._creds = pickle.load(token)
+
+    def _obtain_creds(self):
+        self._load_cached_creds()
+
+        if not self._creds or not self._creds.valid:
+            if self._creds and self._creds.expired and self._creds.refresh_token:
+                self._creds.refresh(Request())
+            else:
+                flow = InstalledAppFlow.from_client_secrets_file(self._credentials_json, SCOPES)
+                creds = flow.run_local_server(port=0)
+                # Save the credentials for the next run
+                if not self._credentials_pickle:
+                    self._credentials_pickle = "token.pickle"
+                with open(self._credentials_pickle, "wb") as token:
+                    pickle.dump(creds, token)
+
+    def _get_service(self):
+        self._obtain_creds()
+        service = build("admin", "directory_v1", credentials=self._creds)
+        return service
+
+    def _get_resource(self, name: str):
+        service = self._get_service()
+        return getattr(service, name)
+
+    @backoff.on_exception(backoff.expo, GoogleApiHttpError, max_tries=7)
+    def get(self, name: str, params: Dict = None):
+        resource = self._get_resource(name)
+        response = resource().list(**params).execute()
+        return response
+
+
+class StreamAPI(ABC):
+    max_results = 10
+
+    def __init__(self, api: API, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._api = api
+
+    def _api_get(self, resource: str, params: Dict = None):
+        return self._api.get(resource, params=params)
+
+    @abstractmethod
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        """Iterate over entities"""
+
+    @abstractmethod
+    def process_response(self, response: Dict) -> Iterator[dict]:
+        """Process Google Directory API response"""
+
+    def read(self, getter: Callable, params: Dict = None) -> Iterator:
+        """Read using getter"""
+        params = params or {}
+        params["maxResults"] = self.max_results
+        while True:
+            batch = getter(params={**params})
+            yield from self.process_response(batch)
+
+            if "nextPageToken" in batch:
+                params["pageToken"] = batch["nextPageToken"]
+            else:
+                break
+
+
+class UsersAPI(StreamAPI):
+    def process_response(self, response: Dict) -> Iterator[dict]:
+        return response["users"]
+
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        params = {"customer": "my_customer", "orderBy": "email"}
```

Comment:
We should add a comment to make this clear 
```suggestion
        # my_customer is an alias for the ID of the user providing credentials
        params = {"customer": "my_customer", "orderBy": "email"}
```